### PR TITLE
OCPBUGS-4704: Include already compliant polices for pre-caching

### PR DESCRIPTION
**Pull request by jc-rh**
Also added a kuttl test with the catalog source policy and one of the subscription policy being compliant prior to the upgrade

### CODE REVIEW

File: controllers/clustergroupupgrade_controller.go [hunk 1]
```diff
@@ -991,6 +992,7 @@ func (r *ClusterGroupUpgradeReconciler) doManagedPoliciesExist(
 
 				if len(clustersNonCompliantWithPolicy) == 0 {
 					managedPoliciesCompliantBeforeUpgrade = append(managedPoliciesCompliantBeforeUpgrade, foundPolicy.GetName())
+					managedPoliciesInfo.compliantPolicies = append(managedPoliciesInfo.compliantPolicies, foundPolicy)
```
Comment (pixelsoccupied) [inline controllers/clustergroupupgrade_controller.go:995]: I wonder this would benefit from moving into a more general place. This block is only called if `filterNonCompliantPolicies` is set to true. i.e a dedicated function which is always called to just grabs all policies needed for precaching operator?

Also  can `managedPoliciesCompliantBeforeUpgrade` and `managedPoliciesInfo.compliantPolicies` be combined into into one? One with just the names and another with the full objects? I think it better to merge them to reduce the complexity.

File: controllers/monitoring.go [hunk 2]
```diff
@@ -106,8 +106,7 @@ func (r *ClusterGroupUpgradeReconciler) getMonitoredObjects(managedPolicy *unstr
 			// Filter only Subscription templates.
 			if !isMonitoredObjectType(kind) {
 				r.Log.Info(
-					"[getPolicyContent] Policy spec.policy-templates.objectDefinition.spec.object-templates.kind does not need to be monitored",
-					"policyName", managedPolicyName)
+					"[getPolicyContent] Object does not need to be monitored", "policyName", managedPolicyName, "kind", kind)
```
Comment (pixelsoccupied) [inline controllers/monitoring.go:109]: Is this related to the PR? But regardless, can we be a bit more explicit as why it doesn't need to be monitored? `Object does not need to be monitored because...`

File: deploy/upgrades/pre-caching/patch-mcv-status-processing-true.sh [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+curl -k -s -X PATCH -H "Accept: application/json, */*" \
+-H "Content-Type: application/merge-patch+json" \
+http://localhost:8001/apis/view.open-cluster-management.io/v1beta1/namespaces/spoke5/managedclusterviews/view-precache-cluster-role-binding/status \
+--data '{"status":{"conditions":[{"lastTransitionTime":"2022-01-28T17:57:00Z","message":"found","reason":"GetResourceProcessing","status":"True","type":"Processing"}]}}'
+
+curl -k -s -X PATCH -H "Accept: application/json, */*" \
+-H "Content-Type: application/merge-patch+json" \
+http://localhost:8001/apis/view.open-cluster-management.io/v1beta1/namespaces/spoke6/managedclusterviews/view-precache-cluster-role-binding/status \
+--data '{"status":{"conditions":[{"lastTransitionTime":"2022-01-28T17:57:00Z","message":"found","reason":"GetResourceProcessing","status":"True","type":"Processing"}]}}'
```
Comment (pixelsoccupied) [inline deploy/upgrades/pre-caching/patch-mcv-status-processing-true.sh:61]: new line please

File: tests/kuttl/tests/pre-caching/01-pre-caching-starting.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+kind: TestStep
+
+commands:
+  # Patch the inform policies to reflect the compliance status Compliant for spoke1.
+  - command: ../../../../deploy/upgrades/pre-caching/patch-ns-mcv-status-processing-false.sh
+    ignoreFailure: false
+  # Force two quick reconcile calls
+  - command: oc --namespace=default patch clustergroupupgrade.ran.openshift.io/cgu --patch '{"spec":{"remediationStrategy":{"timeout":241}}}' --type=merge
+
+  - command: oc --namespace=default patch clustergroupupgrade.ran.openshift.io/cgu --patch '{"spec":{"remediationStrategy":{"timeout":240}}}' --type=merge
```
Comment (pixelsoccupied) [inline tests/kuttl/tests/pre-caching/01-pre-caching-starting.yaml:11]: new line please

File: tests/kuttl/tests/pre-caching/02-pre-caching-starting-deps-ready.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+apiVersion: kuttl.dev/v1beta1
+kind: TestStep
+
+commands:
+  # Patch the inform policies to reflect the compliance status Compliant for spoke1.
+  - command: ../../../../deploy/upgrades/pre-caching/patch-mcv-status-processing-true.sh
+    ignoreFailure: false
+  # Force a quick reconcile
+  - command: oc --namespace=default patch clustergroupupgrade.ran.openshift.io/cgu --patch '{"spec":{"remediationStrategy":{"timeout":241}}}' --type=merge
```
Comment (pixelsoccupied) [inline tests/kuttl/tests/pre-caching/02-pre-caching-starting-deps-ready.yaml:9]: new line please

File: tests/kuttl/tests/pre-caching/03-pre-caching-starting-job-view.yaml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+apiVersion: kuttl.dev/v1beta1
+kind: TestStep
+
+commands:
+  # Patch the inform policies to reflect the compliance status Compliant for spoke1.
+  - command: ../../../../deploy/upgrades/pre-caching/patch-job-mcv-status-processing-false.sh
+    ignoreFailure: false
+  # Force a quick reconcile
+  - command: oc --namespace=default patch clustergroupupgrade.ran.openshift.io/cgu --patch '{"spec":{"remediationStrategy":{"timeout":240}}}' --type=merge
```
Comment (pixelsoccupied) [inline tests/kuttl/tests/pre-caching/03-pre-caching-starting-job-view.yaml:9]: new line please

File: tests/kuttl/tests/pre-caching/04-pre-caching-starting-job-active.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+apiVersion: kuttl.dev/v1beta1
+kind: TestStep
+
+commands:
+  # Patch the inform policies to reflect the compliance status Compliant for spoke1.
+  - command: ../../../../deploy/upgrades/pre-caching/patch-job-mcv-status-processing-true.sh
+    ignoreFailure: false
+  # Force a quick reconcile
+  - command: oc --namespace=default patch clustergroupupgrade.ran.openshift.io/cgu --patch '{"spec":{"remediationStrategy":{"timeout":241}}}' --type=merge
```
Comment (pixelsoccupied) [inline tests/kuttl/tests/pre-caching/04-pre-caching-starting-job-active.yaml:9]: new line please

File: tests/kuttl/tests/pre-caching/05-pre-caching-active-job-succeeded.yaml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+apiVersion: kuttl.dev/v1beta1
+kind: TestStep
+
+commands:
+  # Patch the inform policies to reflect the compliance status Compliant for spoke1.
+  - command: ../../../../deploy/upgrades/pre-caching/patch-job-mcv-status-succeeded.sh
+    ignoreFailure: false
+  # Force a quick reconcile
+  - command: oc --namespace=default patch clustergroupupgrade.ran.openshift.io/cgu --patch '{"spec":{"remediationStrategy":{"timeout":240}}}' --type=merge
```
Comment (pixelsoccupied) [inline tests/kuttl/tests/pre-caching/05-pre-caching-active-job-succeeded.yaml:9]: new line please

File: controllers/monitoring.go (hunk 2 above)
Comment (jc-rh) [reply controllers/monitoring.go:109]: Right, this is not related. Just noticed that it wasn't even a readable sentence. Right now we only monitor subscriptions. We may start monitoring other kinds in the future. That's why I added the kind to the log. I am actually ok with removing the log altogether if you find it confusing. It was there before I joined this project. I think it has served its purposes during the early days of subscription/install plan related development.

File: controllers/clustergroupupgrade_controller.go (hunk 1 above)
Comment (jc-rh) [reply controllers/clustergroupupgrade_controller.go:995]: One is a string array as part of CGU status. So it's part of the API now. The other one is the real policy object that's local to this reconcile.

File: controllers/monitoring.go (hunk 2 above)
Comment (pixelsoccupied) [reply controllers/monitoring.go:109]: Removal sounds good with me.

File: controllers/clustergroupupgrade_controller.go (hunk 1 above)
Comment (pixelsoccupied) [reply controllers/clustergroupupgrade_controller.go:995]: Gotcha. 

So my concern here is that you're populating `managedPoliciesInfo.compliantPolicies` in a func `doManagedPoliciesExist`. The name implies that it's a general func that can be called multiple times by whoever needs to verify that if policies exists. And if that's the case `managedPoliciesInfo.compliantPolicies` maybe populated as many times as well. Which is why I think it's better move this line and populate it in a more explicit way.

File: deploy/upgrades/pre-caching/patch-mcv-status-processing-true.sh (hunk 3 above)
Comment (jc-rh) [reply deploy/upgrades/pre-caching/patch-mcv-status-processing-true.sh:61]: how do you break a line in the middle of an argument string?

Comment (pixelsoccupied) [reply deploy/upgrades/pre-caching/patch-mcv-status-processing-true.sh:61]: Sorry I meant new line at the end of the file

**Comment by pixelsoccupied** [comment]
New change lgtm -- but GitHub still complaining about the missing the end of line. Can you check what's happening there? 

Screen Shot 2022-12-16 at 9 13 47 AM

**Comment by jc-rh** [comment]
> New change lgtm -- but GitHub still complaining about the missing the end of line. Can you check what's happening there?
> 
> Screen Shot 2022-12-16 at 9 13 47 AM

Last line wasn't empty. Good now.

File: tests/kuttl/tests/pre-caching/02-pre-caching-starting-deps-ready.yaml [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+commands:
+  # Patch the inform policies to reflect the compliance status Compliant for spoke1.
+  - command: ../../../../deploy/upgrades/pre-caching/patch-mcv-status-processing-true.sh
+    ignoreFailure: false
+  # Force a quick reconcile
+  - command: >
+      oc --namespace=default patch clustergroupupgrade.ran.openshift.io/cgu
+        --patch '{"spec":{"remediationStrategy":{"timeout":241}}}' --type=merge
+
```
Comment (pixelsoccupied) [inline tests/kuttl/tests/pre-caching/02-pre-caching-starting-deps-ready.yaml:12]: No new at line at end of file

File: tests/kuttl/tests/pre-caching/03-pre-caching-starting-job-view.yaml [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+commands:
+  # Patch the inform policies to reflect the compliance status Compliant for spoke1.
+  - command: ../../../../deploy/upgrades/pre-caching/patch-job-mcv-status-processing-false.sh
+    ignoreFailure: false
+  # Force a quick reconcile
+  - command: >
+      oc --namespace=default patch clustergroupupgrade.ran.openshift.io/cgu 
+        --patch '{"spec":{"remediationStrategy":{"timeout":240}}}' --type=merge
+
```
Comment (pixelsoccupied) [inline tests/kuttl/tests/pre-caching/03-pre-caching-starting-job-view.yaml:12]: No new at line at end of file

File: tests/kuttl/tests/pre-caching/04-pre-caching-starting-job-active.yaml [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+commands:
+  # Patch the inform policies to reflect the compliance status Compliant for spoke1.
+  - command: ../../../../deploy/upgrades/pre-caching/patch-job-mcv-status-processing-true.sh
+    ignoreFailure: false
+  # Force a quick reconcile
+  - command: >
+      oc --namespace=default patch clustergroupupgrade.ran.openshift.io/cgu 
+        --patch '{"spec":{"remediationStrategy":{"timeout":241}}}' --type=merge
+
```
Comment (pixelsoccupied) [inline tests/kuttl/tests/pre-caching/04-pre-caching-starting-job-active.yaml:12]: No new at line at end of file

File: tests/kuttl/tests/pre-caching/05-pre-caching-active-job-succeeded.yaml [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+commands:
+  # Patch the inform policies to reflect the compliance status Compliant for spoke1.
+  - command: ../../../../deploy/upgrades/pre-caching/patch-job-mcv-status-succeeded.sh
+    ignoreFailure: false
+  # Force a quick reconcile
+  - command: >
+      oc --namespace=default patch clustergroupupgrade.ran.openshift.io/cgu 
+        --patch '{"spec":{"remediationStrategy":{"timeout":240}}}' --type=merge
+
```
Comment (pixelsoccupied) [inline tests/kuttl/tests/pre-caching/05-pre-caching-active-job-succeeded.yaml:12]: No new at line at end of file

File: controllers/clustergroupupgrade_controller.go (hunk 1 above)
Comment (jc-rh) [reply controllers/clustergroupupgrade_controller.go]: Not following. 'managedPoliciesInfo' is supposed to be recreated in each call and returned.

**Review by pixelsoccupied — COMMENTED** [review]
Just couple of more end of file line warnings to squash

File: controllers/clustergroupupgrade_controller.go (hunk 1 above)
Comment (pixelsoccupied) [reply controllers/clustergroupupgrade_controller.go]: oh then all good with this.
